### ISSUE TITLE
Integration with ckeditor, isnt showing the images

### ISSUE THREAD

**Issue by elisbresciani** [mileszs/wicked_pdf#698]
So, I load the images from a ckeditor, and when I render the html, it does not show, no error appears, just a blank square, instead of the image, I tried raw, html_safe, but it does not solve, the text I put is showing normally, does anyone have any idea why this happens?

![image](https://user-images.githubusercontent.com/26804018/31889194-ae1084e8-b7dd-11e7-85c7-1ba09d946269.png)

in my view:
```
<h6><b>1. Objetivo:</b></h6>
            <p><%= raw @pop.objective %></p>
```
my controller:
```
  def show
    # redirect_to :index_pdf
    # @pop = Pop.friendly.find(params[:id])
    @pop = Pop.find(params[:id])
   pdf = render pdf: @pop.title,
           :layout => false,
           template: "pops/index_pdf.html.erb",
           locals: {:pop => @pop},
           header: { 
                      right: 'Pág. [page] de [topage].' 
                    }
                    # send_file(pdf, :filename => @pop.title+'.pdf')
                    # baixar automaticamente
                    # precisa fazer 
  end
```
my model:
`has_attached_file :image, styles: { medium: "300x300>" }, default_url: "/images/default.png"`

**Comment by unixmonkey** [comment]
If you were to render that with `show_as_html: true` and view or inspect the source, you'll probably see your calls to `whatever_object.url` translate into relative-paths like `/whatevers/1/thing.png`, 

But for `wicked_pdf` to work, these need to be fully qualified urls or paths like `https://yourdomain.com/whatevers/1/thing.png` (or `file:///path/to/yourapp/assets/images/thing.png` which is what `wicked_pdf_image_tag` does), because the html page is rendered as-is, and passed to `wkhtmltopdf` under-the-hood to render in a headless web browser.

When using wicked_pdf as a middleware, it adds the host automatically with this code:

```ruby
root = WickedPdf.config[:root_url] || "#{env['rack.url_scheme']}://#{env['HTTP_HOST']}/"
body.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + root + '\3\2')
```

So you could preprocess your template like that before passing to `render pdf:`, or, you can modify your calls to `whatever.url` with something like

```ruby
URI.join(ActionController::Base.asset_host, whatever.url)
```

given you have set asset_host somewhere like this:
```
config.action_controller.asset_host = 'https://yourdomain.com'
```

Let me know how it goes!

**Comment by elisbresciani** [comment]
Ok, maybe i have undertand one thing and you said another, i dont have the image, the user will gonna put the image in a text area(ckeditor), and  gonna save in public folder. im using a integration of ckeditor and paperclip . Sorry but i dont understand, im new in rails, can you explain better? And im not using the show_as_html: true. And one month ago was working perfecty, and now... I dont have any idea, i modifyed many things in my project,  maybe one of that thing "broke" the images.

**Comment by unixmonkey** [comment]
I'd have to see the HTML source to see exactly what's going on, but I think you could do something like this:

```ruby
<%= raw @pop.objective.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://yourdomain.com' + '\3\2') %>
```

to transform relative-urls in the source, to absolute URLs.

**Comment by elisbresciani** [comment]
You want see the code? Its ok, i tryed this:
`<%= raw @pop.objective.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://yourdomain.com' + '\3\2') %>
`
but, gives me a log error: 
 
```
F, [2017-10-23T15:05:38.998015 #4968] FATAL -- : [7b26097d-c32c-4b4f-944f-164fa30f8cc1] ActionView::Template::Error (undefined method `gsub' for nil:NilClass):
F, [2017-10-23T15:05:38.998015 #4968] FATAL -- : [7b26097d-c32c-4b4f-944f-164fa30f8cc1]     116:           </div>
[7b26097d-c32c-4b4f-944f-164fa30f8cc1]     117:           <div class="espaco">
[7b26097d-c32c-4b4f-944f-164fa30f8cc1]     118:             <h6><b>3. Documentos relacionados:</b></h6>
[7b26097d-c32c-4b4f-944f-164fa30f8cc1]     119:             <p><%= raw @pop.related_documents.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://192.168.0.203:3000' + '\3\2') %> </p>
[7b26097d-c32c-4b4f-944f-164fa30f8cc1]     120:           </div>
[7b26097d-c32c-4b4f-944f-164fa30f8cc1]     121:           <div class="espaco">
[7b26097d-c32c-4b4f-944f-164fa30f8cc1]     122:             <h6><b>4.  Responsabilidade:</b></h6>
F, [2017-10-23T15:05:38.998015 #4968] FATAL -- : [7b26097d-c32c-4b4f-944f-164fa30f8cc1]   
F, [2017-10-23T15:05:38.998015 #4968] FATAL -- : [7b26097d-c32c-4b4f-944f-164fa30f8cc1] app/views/pops/index_pdf.html.erb:119:in `_app_views_pops_index_pdf_html_erb___137185905_56487696'
[7b26097d-c32c-4b4f-944f-164fa30f8cc1] app/controllers/pops_controller.rb:14:in `show'

```
rails is in production mode...

my view:
```
<!DOCTYPE html>
<html>
<head>
  <!-- colocar no padrao de codificação ele deve estar pegando outro entao é necessario colocar o utf-8 -->
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<!-- Última versão CSS compilada e minificada / bootstrap-->
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">

    <%= javascript_include_tag "http://code.jquery.com/jquery-1.10.0.min.js" %>
    <%= javascript_include_tag "http://code.jquery.com/ui/1.10.3/jquery-ui.min.js" %>


      <title><%= @pop.title  %> + <%= @pop.versions.size %></title>
      <style type="text/css">
        
                        div.container {
                            width: 100%;
                            
                        }
                        nav {
                            float: left;
                            max-width: 160px;
                            margin: 0;
                            padding: 1em;
                        }
                        nav ul {
                            list-style-type: none;
                            padding: 0;
                        }
                        nav ul a {
                            text-decoration: none;
                        }
                        article {
                            align: center;
                            padding: 1em;
                            overflow: hidden;
                            
                        }
                        p  {
                          margin-left: 10px;
                        }
                        div.baixo {

                            padding: 2em;
                        }
                        div.foot {
                            padding: 8em;
                        }
                        div.espaco {
                            padding: 1em;
                        }
                        b.baixo {
                            padding: 15em;
                        }
                        article {
                            padding: 1em;
                            
                        }
                        #ladodireito {

                        padding-top: 0px;
                        padding-right: 0px;
                        padding-left: 150px;
                        padding-bottom: 0px;
                        }
                        div.tamanho {
                            font-size: larger;
                        }
                        b.baixo-baixo{
                          margin-top: 5em;
                          text-align: center;
                        }
                       /*  image_tag.tamanho{
                          width: 120px,
                          height: 90px
                        } */
      </style>
      <script type="text/javascript">
          var headerCheck = function() {
            var x=document.location.search.substring(1).split('&');
            for (var i in x) {
              if(x[i] == "page=1")
                document.getElementsByClassName("headerContent")[0].style.display = "block";
             }
           }();
      </script>
</head>
<body>
  <article>
     <header>
        <table class="table "  >
          <tr>
            <td rowspan="2" colspan="2"  align="center"><%= wicked_pdf_image_tag 'logo.jpg', :width => "100",  :height=>"70" %></td>
            <td rowspan="2" colspan="4" align="center" > <b class="baixo"><h4><b>Procedimento Operacional Padrão</b></h4></b></td>
            <td><b>Doc:</b>   <%= @pop.doc %></td>
            <td colspan="2" rowspan="2" align="center"  class="headerContent" style="display:none">Page <span class="page"></span> of <span class="topage"></span> </td>
          </tr>
          <tr>
            <td><b>Version:</b> <%= @pop.versions.size %></td>
          </tr>
          <tr>
            <td colspan="6" ><center><h5><b id="ladodireito"><%= @pop.title  %></b></h5></center></td>
            <td colspan="2" ><b>  Date:</b> <%= @pop.created_at.to_date  %> </td>
          </tr>
        </table>
      </header>
      <div class="tamanho">
        <div class="baixo" >
          <div class="espaco">
            <h6><b>1. Objetivo:</b></h6>
             <p><%= raw @pop.objective.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://192.168.0.203:3000' + '\3\2') %> </p>
          </div>
          <div class="espaco">
            <h6><b>2. Definições:</b></h6>
            <p><%= raw @pop.settings.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://192.168.0.203:3000' + '\3\2') %> </p>
          </div>
          <div class="espaco">
            <h6><b>3. Documentos relacionados:</b></h6>
            <p><%= raw @pop.related_documents.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://192.168.0.203:3000' + '\3\2') %> </p>
          </div>
          <div class="espaco">
            <h6><b>4.  Responsabilidade:</b></h6>
            <p><%= raw @pop.responsability.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://192.168.0.203:3000' + '\3\2') %> </p>
          </div>
          <div class="espaco">
            <h6><b>5.  Material necessário:</b></h6>
            <p><%= raw @pop.material_needed.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://192.168.0.203:3000' + '\3\2') %> </p>
          </div>
          <div class="espaco">

          <div class="espaco">
            <h6><b>7.  Medidas preventivas:</b></h6>
            <p><%= raw @pop.preventive_measures.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://192.168.0.203:3000' + '\3\2') %> </p>
          </div>
          <div class="espaco">
            <h6><b>8.  Ações corretivas:</b></h6>
            <p><%= raw @pop.corrective_actions.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://192.168.0.203:3000' + '\3\2') %> </p>
          </div>
          
        </div>
      </div>
     <div class="foot">
        <table class="table" >
          <tr>
            <td>
            </td>
            <td>
              <strong>Name</strong>
            </td>
            <td>
              <strong>Data de Aprovação </strong>
            </td>
          </tr>
          <tr>
            <td>
              <strong>Verificador (a) 1 </strong>
            </td>
            <td>
              <%= @pop.checker %>
            </td>
            <td>
              <%= @pop.approval_date_1 %>
            </td>
          </tr>
          <tr>
            <td>
              <strong>Verificador (a) 2 </strong>
            </td>
            <td>
              <%= @pop.checker_2 %>
            </td>
            <td>
              <%= @pop.approval_date_2 %>
            </td>
          </tr>
          <tr>
            <td>
              <strong>Aprovador (a)  </strong>
            </td>
            <td>
              <%= @pop.approver %>
            </td>
            <td>
              <%= @pop.approval_date_3 %>
            </td>
          </tr>
        </table>
    </div>
  </article>
</body>
</html>

```

**Comment by unixmonkey** [comment]
Looks like `@pop.related_documents` is nil for that record. You could add `.to_s` before everything you added the `.gsub` to, like:

```ruby
<%= raw @pop.related_documents.to_s.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://yourdomain.com' + '\3\2') %>
```

This will convert `nil` to the empty string `""`, and `gsub` will no longer blow up for that reason.

And since you are using it all over, it might be helpful to convert into a helper, something like this:

__app/helpers/application_helper.rb__
```ruby
def with_abs_path(htmlstring)
  htmlstring.to_s.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://yourdomain.com' + '\3\2')
end
```

and use like this in your views:

```erb
<%= raw with_abs_path(@pop.related_documents) %>
```

**Comment by elisbresciani** [comment]
i did like you said, but i back to the start, where in local of images appears a blank square. :(
![image](https://user-images.githubusercontent.com/26804018/31908296-218b9926-b815-11e7-883c-c400d9db3406.png)

**Comment by unixmonkey** [comment]
Can you paste the HTML source output of `@pop.related_documents` or some property with an attached image?

**Comment by elisbresciani** [comment]
I'll send all the steps I follow to put an image
To include
![image](https://user-images.githubusercontent.com/26804018/31937533-8b4354a2-b893-11e7-94d1-200485eab53c.png)
included
![image](https://user-images.githubusercontent.com/26804018/31937575-aa3c903a-b893-11e7-95ac-a7332ff47281.png)
in show of rails_admin
![image](https://user-images.githubusercontent.com/26804018/31937599-c61b74e2-b893-11e7-8de3-699f19d77936.png)
field in pdf 
![image](https://user-images.githubusercontent.com/26804018/31937618-d62edf40-b893-11e7-95b4-70728b401302.png)

**Comment by elisbresciani** [comment]
Any idea why this is happening? Any help is helpful, please.

**Comment by unixmonkey** [comment]
Have you tried using wicked_pdf [as a rack middleware](https://github.com/mileszs/wicked_pdf#rack-middleware)?

**Comment by unixmonkey** [comment]
By the way, looking at your above screenshot steps, are the images not showing on normal HTML pages either? You'll need to address that before there are any hopes of displaying them in a PDF.

**Comment by elisbresciani** [comment]
It is showing, let me edit and put another example.

**Comment by elisbresciani** [comment]
It is only in the pdf that is not showing, in others this normal.

**Comment by elisbresciani** [comment]
I do not know how rails_admin will behave.

**Comment by elisbresciani** [comment]
Soooo????? Any idea?

**Comment by elisbresciani** [comment]
Hello, how are you? On the problem, I was able to identify, it is, I believe, in the part of wkhtml, I moved the controller, I left it only being html and the images appear, but when it goes to pdf, it does not appear anymore. Please help me, I'm using the tool at work, it's already on the intranet, this part of the pdf is missing.

**Comment by unixmonkey** [comment]
Can you paste the HTML source output of `@pop.related_documents` or some property with an image in it?

It should output the raw HTML and look something like this:

```ruby
puts @pop.related_documents
<div id="editor"><ol><li>Objectivo:</li></ol><div>teste teste teste teste</div><img src="http://yourdomain.com/images/db_diagram.jpg"></div>
```

I'm especially interested in what the image's `src` attribute is. This could be a fully-qualified path to your app like `http://yourdomain.com/public/images/db_diagram.jpg`, or a file path like `file://Users/elisbresciani/my_app/public/images/db_diagram.jpg`, but cannot be a relative-path like `/public/images/db_diagram.jpg', because the "webpage" that gets converted to a PDF is not being served from the web, so relative references won't work.

**Comment by elisbresciani** [comment]
ok I did the test, and well, it's the relative path, but the odd one I have an older version of my project, I was already integrated, and the relative path works.
![image](https://user-images.githubusercontent.com/26804018/32546805-633698c2-c467-11e7-98e2-bafe0fb64592.png)

**Comment by unixmonkey** [comment]
Does it show a full path to the image if you use the helper method I described above?

```erb
<%= raw with_abs_path(@pop.related_documents) %>
```

**Comment by elisbresciani** [comment]
No, even with that still not showing the image.

**Comment by unixmonkey** [comment]
Please paste the source so we can see the `<img src="">` value.

**Comment by elisbresciani** [comment]
ok, here is the code:

    <!DOCTYPE html>
    <html>
      <head>
    <title>Net Pluma</title>
    <meta name="csrf-param" content="authenticity_token" />
    <meta name="csrf-token" content="HWziCsHbmBhd2Mxoc+uEVTqzESf/JlClzfHCS035KjN42sYRus9SDMqARRoAI3xdlLvIAY6ktwoblT0EGFecig==" />
    <link rel="stylesheet" media="all" href="/assets/application-1f4edc2b7069dfbe72faec62c6157876db6cc8ebaafe78c5d5c15ffd5a719dc3.css" data-turbolinks-track="reload" />
    <script src="/assets/application-9b0234518cae1e697b32e04127ee38e1a1b09c0d98037815fda26ea447798170.js" data-turbolinks-track="reload"></script>
      </head>
    
      <body>
        
    <!DOCTYPE html>
    <html>
    <head>
      
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
    <script src="/javascripts/ckeditor/ckeditor.js"></script>
    <script src="http://code.jquery.com/jquery-1.10.0.min.js"></script>
    <script src="http://code.jquery.com/ui/1.10.3/jquery-ui.min.js"></script>


      <title>teste1Elis Bresciani</title>
      <style type="text/css">
        
                        div.container {
                            width: 100%;
                            
                        }
                        nav {
                            float: left;
                            max-width: 160px;
                            margin: 0;
                            padding: 1em;
                        }
                        nav ul {
                            list-style-type: none;
                            padding: 0;
                        }
                        nav ul a {
                            text-decoration: none;
                        }
                        article {
                            align: center;
                            padding: 1em;
                            overflow: hidden;
                            
                        }
                        p  {
                          margin-left: 10px;
                        }
                        div.baixo {

                            padding: 2em;
                        }
                        div.foot {
                            padding: 8em;
                        }
                        div.espaco {
                            padding: 1em;
                        }
                        b.baixo {
                            padding: 15em;
                        }
                        article {
                            padding: 1em;
                            
                        }
                        #ladodireito {

                        padding-top: 0px;
                        padding-right: 0px;
                        padding-left: 150px;
                        padding-bottom: 0px;
                        }
                        div.tamanho {
                            font-size: larger;
                        }
                        b.baixo-baixo{
                          margin-top: 5em;
                          text-align: center;
                        }
                       /*  image_tag.tamanho{
                          width: 120px,
                          height: 90px
                        } */
      </style>
      <script type="text/javascript">
          var headerCheck = function() {
            var x=document.location.search.substring(1).split('&');
            for (var i in x) {
              if(x[i] == "page=1")
                document.getElementsByClassName("headerContent")[0].style.display = "block";
             }
           }();
      </script>
    </head>
    <body>
      <article>
         <header>
            <table class="table "  >
              <tr>
                <td rowspan="2" colspan="2"  align="center"><img width="100" height="70" src="file:///C:/Arquivos/proj/tags/tags_v2/public/assets/logo-36ad01480e4610ed11304c0ab2795d638b4aeff5a940339bfc9e1c6e1c87db15.jpg" alt="Logo" /></td>
            <td rowspan="2" colspan="4" align="center" > <b class="baixo"><h4><b>Procedimento Operacional Padrão</b></h4></b></td>
            <td><b>Doc:</b>   </td>
            <td colspan="2" rowspan="2" align="center"  class="headerContent" style="display:none">Page<span class="page"></span> of <span class="topage"></span> </td>
          </tr>
          <tr>
            <td><b>Version:</b> 1</td>
          </tr>
          <tr>
            <td colspan="6" ><center><h5><b id="ladodireito">teste</b></h5></center></td>
            <td colspan="2" ><b>  Date:</b> 2017-11-10 </td>
          </tr>
        </table>
      </header>
      <div class="tamanho">
        <div class="baixo" >
          <div class="espaco">
            <h6><b>1. Objetivo:</b></h6>
            <p>teste teste&nbsp;<br />
    <img alt="" src="http://182.168.0.203:3000ckeditor_assets/pictures/385/content_campo.png" style="width: 800px; height: 212px;" /></p>
              </div>
    

      <div class="espaco">
            <h6><b>2. Definições:</b></h6>
            <p></p>
          </div>

          <div class="espaco">
            <h6><b>4.  Responsabilidade:</b></h6>
            <p></p>
          </div>
          <div class="espaco">
            <h6><b>5.  Material necessário:</b></h6>
            <p></p>
          </div>
          <div class="espaco">
            <h6><b>6.  Descrição:</b></h6>
            <p></p>
          </div>
          <div class="espaco">
            <h6><b>7.  Medidas preventivas:</b></h6>
            <p></p>
          </div>
          <div class="espaco">
            <h6><b>8.  Ações corretivas:</b></h6>
            <p></p>
          </div>
        </div>
      </div>
     <div class="foot">
        <table class="table" >
          <tr>
            <td>
            </td>
            <td>
              <strong>Name</strong>
            </td>
            <td>
              <strong>Data de Aprovação </strong>
            </td>
          </tr>
          <tr>
            <td>
              <strong>Verificador (a) 1 </strong>
            </td>
            <td>
              
            </td>
            <td>
              
            </td>
          </tr>
          <tr>
            <td>
              <strong>Verificador (a) 2 </strong>
            </td>
            <td>
              
            </td>
            <td>
              
            </td>
          </tr>
          <tr>
            <td>
              <strong>Aprovador (a)  </strong>
            </td>
            <td>
              
            </td>
            <td>
              
            </td>
          </tr>
        </table>
    </div>
      </article>
    </body>
    </html>
    
    
      </body>
    </html>

**Comment by unixmonkey** [comment]
`http://182.168.0.203:3000ckeditor_assets/pictures/385/content_campo.png`, is missing a forward-slash after the 3000.

`/assets/application-9b0234518cae1e697b32e04127ee38e1a1b09c0d98037815fda26ea447798170.js` and 
`/javascripts/ckeditor/ckeditor.js` are both using a relative path. Add them to your asset precompile list, and use `wicked_pdf_javascript_include_tag`, or offload to a CDN.

```ruby
config.assets.precompile += ['application.js', 'ckeditor/ckeditor.js']
```

`logo-36ad01480e4610ed11304c0ab2795d638b4aeff5a940339bfc9e1c6e1c87db15.jpg` is working perfectly from what I can tell. No change should be necessary here.

**Comment by elisbresciani** [comment]
I was able to solve the problem. I took part of his response and implemented, this after deploy on the network for a hot ip, I left the ckeditor saving the images normally, and when he searches to generate the pdf I put the part of it to include the cold ip in the links and voila . It worked. thank you very much for your help!

> app/helpers/application_helper.rb

```
def with_abs_path(htmlstring)
  htmlstring.to_s.gsub(/(href|src)=(['"])\/([^\"']*|[^"']*)['"]/, '\1=\2' + 'http://yourdomain.com/' + '\3\2')
end
```
and in my view.
`<%= raw with_abs_path(@pop.related_documents) %>`
